File: erstelle_pdf.py
from reportlab.lib.pagesizes import A4
from reportlab.lib import colors
from reportlab.lib.units import cm
from reportlab.platypus import SimpleDocTemplate, Table, TableStyle, Paragraph, Spacer, HRFlowable
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.lib.enums import TA_CENTER, TA_LEFT, TA_RIGHT

produkte = [
    {
        'name': 'HORI Stelzlager',
        'shop': 'casando.de',
        'hoehe': '32 - 45 mm',
        'belastbarkeit': 'ca. 2.000 kg/m2',
        'fuge': '3,5 mm (fix)',
        'gefaelle': '0 - 5 % = 0 - 3 Grad',
        'oben': 'ca. 100 mm',
        'unten': 'ca. 160 mm',
        'preis': 2.89,
        'lieferung': '12.06. - 15.06.',
        'selbstnivellierend': 'NEIN (manuell)',
    },
    {
        'name': 'ceratrends EasyPro',
        'shop': 'ceratrends.com',
        'hoehe': '28 - 36 mm',
        'belastbarkeit': 'ca. 2.000 kg/m2',
        'fuge': '2 mm oder 4 mm',
        'gefaelle': '0 - 5 % = 0 - 3 Grad',
        'oben': 'k. A.',
        'unten': 'k. A.',
        'preis': 3.29,
        'lieferung': '16.06. - 17.06.',
        'selbstnivellierend': 'JA (automatisch)',
    },
    {
        'name': 'Eterno Ivica SE0',
        'shop': 'markenfliesen24.de',
        'hoehe': '28 - 38 mm',
        'belastbarkeit': 'k. A.',
        'fuge': '2 mm oder 3 mm',
        'gefaelle': '0 - 5 % = 0 - 3 Grad',
        'oben': '110 mm',
        'unten': '205 mm',
        'preis': 3.56,
        'lieferung': '1 - 3 Tage',
        'selbstnivellierend': 'JA (automatisch)',
    },
]

stueckzahl = 180
reserve_stueck = 18  # 10%
gesamt_stueck = 198

BLAU = colors.HexColor('#1a3a5c')
HELLBLAU = colors.HexColor('#dce8f5')
GRAU = colors.HexColor('#f5f5f5')
DUNKELGRAU = colors.HexColor('#333333')
GRUEN = colors.HexColor('#1a7a3c')
HELLGRUEN = colors.HexColor('#e8f5ec')
ROT = colors.HexColor('#b22222')

doc = SimpleDocTemplate(
    'Stelzlager_Vergleich.pdf',
    pagesize=A4,
    rightMargin=1.5*cm,
    leftMargin=1.5*cm,
    topMargin=2*cm,
    bottomMargin=2*cm
)

def ps(name, **kwargs):
    return ParagraphStyle(name, **kwargs)

story = []

# --- TITEL ---
story.append(Paragraph('Stelzlager Vergleich', ps('T', fontSize=18, textColor=BLAU, fontName='Helvetica-Bold', alignment=TA_CENTER, spaceAfter=3)))
story.append(Paragraph('Selbstnivellierende Stelzlager  |  40 m2 Terrasse  |  Stand: 10.06.2025', ps('U', fontSize=9, textColor=DUNKELGRAU, alignment=TA_CENTER, spaceAfter=2)))
story.append(HRFlowable(width='100%', thickness=2, color=BLAU, spaceAfter=14))

# --- TECHNISCHE DATEN ---
story.append(Paragraph('1.  Technische Spezifikationen', ps('S', fontSize=12, textColor=BLAU, fontName='Helvetica-Bold', spaceBefore=4, spaceAfter=7)))

header = [
    Paragraph('Eigenschaft', ps('h', fontSize=8.5, textColor=colors.white, fontName='Helvetica-Bold', alignment=TA_LEFT)),
    Paragraph('HORI\ncasando.de', ps('h', fontSize=8.5, textColor=colors.white, fontName='Helvetica-Bold', alignment=TA_CENTER)),
    Paragraph('EasyPro\nceratrends.com', ps('h', fontSize=8.5, textColor=colors.white, fontName='Helvetica-Bold', alignment=TA_CENTER)),
    Paragraph('Eterno Ivica SE0\nmarkenfliesen24.de', ps('h', fontSize=8.5, textColor=colors.white, fontName='Helvetica-Bold', alignment=TA_CENTER)),
]

def row(label, v1, v2, v3):
    s = ps('c', fontSize=8.5, textColor=DUNKELGRAU, alignment=TA_CENTER)
    sl = ps('cl', fontSize=8.5, textColor=DUNKELGRAU, fontName='Helvetica-Bold')
    return [Paragraph(label, sl), Paragraph(v1, s), Paragraph(v2, s), Paragraph(v3, s)]

tech = [
    header,
    row('Hoehenausgleich', '32 - 45 mm', '28 - 36 mm', '28 - 38 mm'),
    row('Selbstnivellierend', 'NEIN (manuell)', 'JA (automatisch)', 'JA (automatisch)'),
    row('Gefalleausgleich', '0 - 5 % = 0 - 3 Grad', '0 - 5 % = 0 - 3 Grad', '0 - 5 % = 0 - 3 Grad'),
    row('Fugenbreite', '3,5 mm (fix)', '2 mm oder 4 mm', '2 mm oder 3 mm'),
    row('Belastbarkeit', 'ca. 2.000 kg/m2', 'ca. 2.000 kg/m2', 'k. A.'),
    row('Auflagef. oben', 'ca. 100 mm', 'k. A.', '110 mm'),
    row('Auflagef. unten (Fuss)', 'ca. 160 mm', 'k. A.', '205 mm'),
    row('Lieferzeit', '12.06. - 15.06.', '16.06. - 17.06.', '1 - 3 Tage'),
    row('Preis / Stuck', '2,89 EUR', '3,29 EUR', '3,56 EUR'),
]

cw = [4.5*cm, 4.3*cm, 4.3*cm, 4.6*cm]
tt = Table(tech, colWidths=cw, repeatRows=1)
ts_tech = [
    ('BACKGROUND', (0,0), (-1,0), BLAU),
    ('BACKGROUND', (0,1), (0,-1), HELLBLAU),
    ('FONTNAME', (0,1), (0,-1), 'Helvetica-Bold'),
    ('BACKGROUND', (1,2), (-1,2), colors.HexColor('#fde8e8')),
    ('TEXTCOLOR', (1,2), (1,2), ROT),
    ('TEXTCOLOR', (2,2), (3,2), GRUEN),
    ('FONTNAME', (1,2), (3,2), 'Helvetica-Bold'),
    ('BACKGROUND', (1,4), (-1,4), GRAU),
    ('BACKGROUND', (1,6), (-1,6), GRAU),
    ('BACKGROUND', (1,8), (-1,8), GRAU),
    ('BACKGROUND', (0,9), (-1,9), colors.HexColor('#fffbe6')),
    ('FONTNAME', (0,9), (-1,9), 'Helvetica-Bold'),
    ('FONTSIZE', (0,9), (-1,9), 9),
    ('GRID', (0,0), (-1,-1), 0.5, colors.HexColor('#cccccc')),
    ('ALIGN', (1,0), (-1,-1), 'CENTER'),
    ('VALIGN', (0,0), (-1,-1), 'MIDDLE'),
    ('TOPPADDING', (0,0), (-1,-1), 5),
    ('BOTTOMPADDING', (0,0), (-1,-1), 5),
    ('LEFTPADDING', (0,0), (-1,-1), 7),
    ('RIGHTPADDING', (0,0), (-1,-1), 7),
]
tt.setStyle(TableStyle(ts_tech))
story.append(tt)
story.append(Spacer(1, 16))

# --- KOSTENBERECHNUNG ---
story.append(Paragraph('2.  Kostenberechnung pro Produkt', ps('S2', fontSize=12, textColor=BLAU, fontName='Helvetica-Bold', spaceBefore=4, spaceAfter=4)))
story.append(Paragraph(
    'Grundbedarf: 180 Stuck  |  10 % Reserve: 18 Stuck  |  Gesamt: 198 Stuck',
    ps('Info', fontSize=9, textColor=DUNKELGRAU, fontName='Helvetica-Oblique', spaceAfter=10)
))

for p in produkte:
    k180 = stueckzahl * p['preis']
    kres = reserve_stueck * p['preis']
    kges = gesamt_stueck * p['preis']

    kopf = [
        Paragraph(p['name'] + '   |   ' + p['shop'], ps('kh', fontSize=9.5, textColor=colors.white, fontName='Helvetica-Bold')),
        '', '', ''
    ]
    col_h = [
        Paragraph('Position', ps('ch', fontSize=8, textColor=BLAU, fontName='Helvetica-Bold')),
        Paragraph('Stuck', ps('ch', fontSize=8, textColor=BLAU, fontName='Helvetica-Bold', alignment=TA_CENTER)),
        Paragraph('Preis / Stuck', ps('ch', fontSize=8, textColor=BLAU, fontName='Helvetica-Bold', alignment=TA_CENTER)),
        Paragraph('Betrag', ps('ch', fontSize=8, textColor=BLAU, fontName='Helvetica-Bold', alignment=TA_RIGHT)),
    ]
    r1 = [
        Paragraph('Grundbedarf', ps('r', fontSize=9, textColor=DUNKELGRAU)),
        Paragraph('180', ps('rc', fontSize=9, textColor=DUNKELGRAU, alignment=TA_CENTER)),
        Paragraph('2,89 EUR' if p['preis'] == 2.89 else ('3,29 EUR' if p['preis'] == 3.29 else '3,56 EUR'), ps('rc', fontSize=9, textColor=DUNKELGRAU, alignment=TA_CENTER)),
        Paragraph(f'{k180:.2f} EUR', ps('rr', fontSize=9, textColor=DUNKELGRAU, alignment=TA_RIGHT)),
    ]
    r2 = [
        Paragraph('10 % Reserve (18 Stuck)', ps('r', fontSize=9, textColor=DUNKELGRAU)),
        Paragraph('18', ps('rc', fontSize=9, textColor=DUNKELGRAU, alignment=TA_CENTER)),
        Paragraph('2,89 EUR' if p['preis'] == 2.89 else ('3,29 EUR' if p['preis'] == 3.29 else '3,56 EUR'), ps('rc', fontSize=9, textColor=colors.HexColor('#888888'), alignment=TA_CENTER)),
        Paragraph(f'{kres:.2f} EUR', ps('rr', fontSize=9, textColor=colors.HexColor('#888888'), alignment=TA_RIGHT)),
    ]
    r3 = [
        Paragraph('<b>GESAMT (198 Stuck)</b>', ps('rg', fontSize=10, fontName='Helvetica-Bold', textColor=DUNKELGRAU)),
        Paragraph('<b>198</b>', ps('rgc', fontSize=10, fontName='Helvetica-Bold', textColor=DUNKELGRAU, alignment=TA_CENTER)),
        Paragraph(f'<b>{p["preis"]:.2f} EUR</b>', ps('rgc2', fontSize=10, fontName='Helvetica-Bold', textColor=DUNKELGRAU, alignment=TA_CENTER)),
        Paragraph(f'<b>{kges:.2f} EUR</b>', ps('rgr', fontSize=12, fontName='Helvetica-Bold', textColor=GRUEN, alignment=TA_RIGHT)),
    ]

    calc_data = [kopf, col_h, r1, r2, r3]
    cw2 = [6.5*cm, 2.3*cm, 3.5*cm, 5.4*cm]
    ct = Table(calc_data, colWidths=cw2)
    ct.setStyle(TableStyle([
        ('BACKGROUND', (0,0), (-1,0), BLAU),
        ('SPAN', (0,0), (-1,0)),
        ('BACKGROUND', (0,1), (-1,1), HELLBLAU),
        ('BACKGROUND', (0,2), (-1,2), colors.white),
        ('BACKGROUND', (0,3), (-1,3), GRAU),
        ('BACKGROUND', (0,4), (-1,4), HELLGRUEN),
        ('LINEABOVE', (0,4), (-1,4), 1.5, GRUEN),
        ('GRID', (0,1), (-1,-1), 0.4, colors.HexColor('#cccccc')),
        ('BOX', (0,0), (-1,-1), 1, BLAU),
        ('VALIGN', (0,0), (-1,-1), 'MIDDLE'),
        ('TOPPADDING', (0,0), (-1,-1), 5),
        ('BOTTOMPADDING', (0,0), (-1,-1), 5),
        ('LEFTPADDING', (0,0), (-1,-1), 8),
        ('RIGHTPADDING', (0,0), (-1,-1), 8),
    ]))
    story.append(ct)
    story.append(Spacer(1, 10))

# --- ZUSAMMENFASSUNG ---
story.append(HRFlowable(width='100%', thickness=1.5, color=BLAU, spaceBefore=6, spaceAfter=8))
story.append(Paragraph('3.  Gesamtuebersicht & Rangliste (198 Stuck inkl. 10 % Reserve)', ps('S3', fontSize=12, textColor=BLAU, fontName='Helvetica-Bold', spaceBefore=4, spaceAfter=8)))

sortiert = sorted(produkte, key=lambda x: x['preis'])
rang_farben = [HELLGRUEN, GRAU, colors.HexColor('#fde8e8')]
rang_label = ['1. Guenstigster', '2. Mittelfeld', '3. Teuerster']

sum_header = [
    Paragraph('Rang', ps('sh', fontSize=9, fontName='Helvetica-Bold', textColor=colors.white, alignment=TA_CENTER)),
    Paragraph('Produkt', ps('sh', fontSize=9, fontName='Helvetica-Bold', textColor=colors.white)),
    Paragraph('Preis/Stk.', ps('sh', fontSize=9, fontName='Helvetica-Bold', textColor=colors.white, alignment=TA_CENTER)),
    Paragraph('Gesamt 198 Stk.', ps('sh', fontSize=9, fontName='Helvetica-Bold', textColor=colors.white, alignment=TA_RIGHT)),
    Paragraph('Selbstnivellierend', ps('sh', fontSize=9, fontName='Helvetica-Bold', textColor=colors.white, alignment=TA_CENTER)),
]
sum_data = [sum_header]
for i, p in enumerate(sortiert):
    g = gesamt_stueck * p['preis']
    sniv_col = GRUEN if p['selbstnivellierend'].startswith('JA') else ROT
    sum_data.append([
        Paragraph(rang_label[i], ps(f'rl{i}', fontSize=8.5, fontName='Helvetica-Bold', alignment=TA_CENTER)),
        Paragraph(f'<b>{p["name"]}</b>', ps(f'pl{i}', fontSize=9, fontName='Helvetica-Bold')),
        Paragraph(f'{p["preis"]:.2f} EUR', ps(f'pr{i}', fontSize=9, alignment=TA_CENTER)),
        Paragraph(f'<b>{g:.2f} EUR</b>', ps(f'gt{i}', fontSize=10, fontName='Helvetica-Bold', alignment=TA_RIGHT)),
        Paragraph(p['selbstnivellierend'], ps(f'sn{i}', fontSize=8.5, textColor=sniv_col, fontName='Helvetica-Bold', alignment=TA_CENTER)),
    ])

scw = [3.0*cm, 4.8*cm, 2.8*cm, 4.2*cm, 3.8*cm]
st = Table(sum_data, colWidths=scw)
sts = [
    ('BACKGROUND', (0,0), (-1,0), BLAU),
    ('GRID', (0,0), (-1,-1), 0.5, colors.HexColor('#cccccc')),
    ('BOX', (0,0), (-1,-1), 1.5, BLAU),
    ('VALIGN', (0,0), (-1,-1), 'MIDDLE'),
    ('TOPPADDING', (0,0), (-1,-1), 6),
    ('BOTTOMPADDING', (0,0), (-1,-1), 6),
    ('LEFTPADDING', (0,0), (-1,-1), 7),
    ('RIGHTPADDING', (0,0), (-1,-1), 7),
]
for i in range(3):
    sts.append(('BACKGROUND', (0, i+1), (-1, i+1), rang_farben[i]))
st.setStyle(TableStyle(sts))
story.append(st)

story.append(Spacer(1, 14))
story.append(HRFlowable(width='100%', thickness=0.5, color=colors.HexColor('#aaaaaa'), spaceAfter=5))
story.append(Paragraph(
    'Hinweis: Alle Produkte decken maximal 5 % Gefaelle aus. '
    'HORI nivelliert NICHT automatisch - manuelles Ausrichten jedes Stelzlagers erforderlich. '
    'Preisstand: 10.06.2025. Alle Angaben ohne Gewahr.',
    ps('foot', fontSize=7.5, textColor=colors.HexColor('#666666'), fontName='Helvetica-Oblique')
))

doc.build(story)
print('PDF erfolgreich erstellt: Stelzlager_Vergleich.pdf')

File: erstelle_pdf.js
const PDFDocument = require('pdfkit');
const fs = require('fs');

const doc = new PDFDocument({ size: 'A4', margin: 40 });
doc.pipe(fs.createWriteStream('Stelzlager_Vergleich.pdf'));

const BLAU = '#1a3a5c';
const HELLBLAU = '#dce8f5';
const GRAU = '#f0f0f0';
const HELLGRUEN = '#e8f5ec';
const GRUEN = '#1a7a3c';
const ROT = '#b22222';
const GELB = '#fffbe6';
const DUNKELGRAU = '#333333';
const ORANGE_BG = '#fff3e0';
const WHITE = '#ffffff';

const W = 515;
const LEFT = 40;
const STUECK = 180;
const AUFSCHLAG = 0.10;

const produkte = [
  {
    name: 'HORI Stelzlager',
    shop: 'casando.de',
    hoehe: '32 - 45 mm',
    belastbarkeit: 'ca. 2.000 kg/m²',
    fuge: '3,5 mm (fix)',
    gefaelle: '0-5 % = 0-3°',
    oben: 'ca. 100 mm',
    unten: 'ca. 160 mm',
    preis: 2.89,
    lieferung: '12.06. - 15.06.',
  },
  {
    name: 'ceratrends EasyPro',
    shop: 'ceratrends.com',
    hoehe: '28 - 36 mm',
    belastbarkeit: 'ca. 2.000 kg/m²',
    fuge: '2 mm oder 4 mm',
    gefaelle: '0-5 % = 0-3°',
    oben: 'k. A.',
    unten: 'k. A.',
    preis: 3.29,
    lieferung: '16.06. - 17.06.',
  },
  {
    name: 'Eterno Ivica SE0',
    shop: 'markenfliesen24.de',
    hoehe: '28 - 38 mm',
    belastbarkeit: 'k. A.',
    fuge: '2 mm oder 3 mm',
    gefaelle: '0-5 % = 0-3°',
    oben: '110 mm',
    unten: '205 mm',
    preis: 3.56,
    lieferung: '1 - 3 Tage',
  },
];

function rect(x, y, w, h, color) {
  doc.save().rect(x, y, w, h).fill(color).restore();
}

function txt(text, x, y, w, opts = {}) {
  const { color = DUNKELGRAU, bold = false, size = 9, align = 'left', italic = false } = opts;
  let font = 'Helvetica';
  if (bold) font = 'Helvetica-Bold';
  if (italic) font = 'Helvetica-Oblique';
  doc.save().font(font).fontSize(size).fillColor(color).text(text, x, y, { width: w, align }).restore();
}

// ============================
// TITEL
// ============================
rect(LEFT, 40, W, 48, BLAU);
txt('Stelzlager Vergleich', LEFT, 52, W, { color: WHITE, bold: true, size: 18, align: 'center' });
txt('Selbstnivellierende Stelzlager  |  40 m²  |  Stand: 10.06.2025', LEFT, 76, W, { color: '#aaccee', size: 9, align: 'center' });

let y = 104;

// ============================
// 1. TECHNISCHE DATEN
// ============================
txt('1.  Technische Spezifikationen', LEFT, y, W, { color: BLAU, bold: true, size: 12 });
y += 16;

const cols = [162, 111, 121, 121];
const colX = [LEFT, LEFT + cols[0], LEFT + cols[0] + cols[1], LEFT + cols[0] + cols[1] + cols[2]];
const rowH = 21;

// Header
rect(LEFT, y, W, rowH, BLAU);
['Eigenschaft', 'HORI  (casando.de)', 'EasyPro  (ceratrends)', 'Eterno Ivica SE0'].forEach((h, i) => {
  txt(h, colX[i] + 5, y + 7, cols[i] - 10, { color: WHITE, bold: true, size: 8, align: i === 0 ? 'left' : 'center' });
});
y += rowH;

// Technische Daten (OHNE Selbstnivellierend)
const techRows = [
  ['Hoehenausgleich',       '32 - 45 mm',       '28 - 36 mm',       '28 - 38 mm'],
  ['Gefalleausgleich',      '0-5 % = 0-3°',     '0-5 % = 0-3°',     '0-5 % = 0-3°'],
  ['Fugenbreite',           '3,5 mm (fix)',      '2 mm oder 4 mm',   '2 mm oder 3 mm'],
  ['Belastbarkeit',         'ca. 2.000 kg/m²',  'ca. 2.000 kg/m²',  'k. A.'],
  ['Auflagef. oben (Kopf)', 'ca. 100 mm',        'k. A.',             '110 mm'],
  ['Auflagef. unten (Fuss)','ca. 160 mm',        'k. A.',             '205 mm'],
  ['Lieferzeit',            '12.06.-15.06.',     '16.06.-17.06.',    '1 - 3 Tage'],
  ['Preis / Stueck',        '2,89 EUR',          '3,29 EUR',         '3,56 EUR'],
];

techRows.forEach((row, ri) => {
  const rowBg = ri % 2 === 0 ? WHITE : GRAU;
  rect(LEFT, y, W, rowH, rowBg);
  rect(LEFT, y, cols[0], rowH, HELLBLAU);
  if (ri === 7) { // Preis-Zeile
    rect(LEFT, y, W, rowH, GELB);
    rect(LEFT, y, cols[0], rowH, HELLBLAU);
  }
  row.forEach((val, ci) => {
    txt(val, colX[ci] + 5, y + 7, cols[ci] - 10, {
      bold: ci === 0 || ri === 7,
      size: ri === 7 ? 9.5 : 8.5,
      align: ci === 0 ? 'left' : 'center',
    });
  });
  doc.save().moveTo(LEFT, y + rowH).lineTo(LEFT + W, y + rowH).lineWidth(0.4).strokeColor('#cccccc').stroke().restore();
  y += rowH;
});
doc.save().rect(LEFT, y - rowH * techRows.length - rowH, W, rowH * (techRows.length + 1)).lineWidth(1).strokeColor(BLAU).stroke().restore();
y += 18;

// ============================
// 2. KOSTENBERECHNUNG
// ============================
txt('2.  Kostenberechnung', LEFT, y, W, { color: BLAU, bold: true, size: 12 });
y += 14;
txt(
  `Berechnung: 180 Stueck  x  Einzelpreis  +  10 % Preisaufschlag`,
  LEFT, y, W, { italic: true, size: 9 }
);
y += 14;

const cw2 = [220, 80, 110, 105];
const cx2 = [LEFT, LEFT + 220, LEFT + 300, LEFT + 410];

produkte.forEach(p => {
  const preisAuf = p.preis * (1 + AUFSCHLAG);
  const basisGes = STUECK * p.preis;
  const aufGes   = STUECK * p.preis * AUFSCHLAG;
  const totalGes = STUECK * preisAuf;

  const fmt = n => n.toFixed(2).replace('.', ',') + ' EUR';
  const fmtU = n => n.toFixed(2).replace('.', ',');

  // Produktkopf
  rect(LEFT, y, W, 22, BLAU);
  txt(`${p.name}   |   ${p.shop}`, LEFT + 10, y + 6, W - 20, { color: WHITE, bold: true, size: 10 });
  y += 22;

  // Spalten-Header
  rect(LEFT, y, W, 19, HELLBLAU);
  ['Position', 'Stueck', 'Preis / Stueck', 'Betrag'].forEach((h, i) => {
    txt(h, cx2[i] + 5, y + 5, cw2[i] - 10, { color: BLAU, bold: true, size: 8, align: i === 0 ? 'left' : 'right' });
  });
  y += 19;

  // Zeile 1: Grundpreis
  rect(LEFT, y, W, 21, WHITE);
  txt('180 Stueck x Einzelpreis', cx2[0] + 5, y + 6, cw2[0] - 10, { size: 9 });
  txt('180',               cx2[1] + 5, y + 6, cw2[1] - 10, { size: 9, align: 'right' });
  txt(fmt(p.preis),        cx2[2] + 5, y + 6, cw2[2] - 10, { size: 9, align: 'right' });
  txt(fmt(basisGes),       cx2[3] + 5, y + 6, cw2[3] - 10, { size: 9, align: 'right' });
  doc.save().moveTo(LEFT, y + 21).lineTo(LEFT + W, y + 21).lineWidth(0.4).strokeColor('#cccccc').stroke().restore();
  y += 21;

  // Zeile 2: 10% Aufschlag
  rect(LEFT, y, W, 21, GRAU);
  txt('+ 10 % Preisaufschlag', cx2[0] + 5, y + 6, cw2[0] - 10, { size: 9 });
  txt('',                     cx2[1] + 5, y + 6, cw2[1] - 10, { size: 9, align: 'right' });
  txt(`+ ${fmtU(p.preis * AUFSCHLAG)} EUR`, cx2[2] + 5, y + 6, cw2[2] - 10, { size: 9, align: 'right', color: '#666666' });
  txt(`+ ${fmt(aufGes)}`,    cx2[3] + 5, y + 6, cw2[3] - 10, { size: 9, align: 'right', color: '#666666' });
  doc.save().moveTo(LEFT, y + 21).lineTo(LEFT + W, y + 21).lineWidth(0.4).strokeColor('#cccccc').stroke().restore();
  y += 21;

  // Gesamt-Zeile
  rect(LEFT, y, W, 24, HELLGRUEN);
  doc.save().moveTo(LEFT, y).lineTo(LEFT + W, y).lineWidth(1.5).strokeColor(GRUEN).stroke().restore();
  txt('GESAMT (180 Stueck)', cx2[0] + 5, y + 7, cw2[0] - 10, { bold: true, size: 10 });
  txt('180',                 cx2[1] + 5, y + 7, cw2[1] - 10, { bold: true, size: 10, align: 'right' });
  txt(fmt(preisAuf) + ' *', cx2[2] + 5, y + 7, cw2[2] - 10, { bold: true, size: 9, align: 'right' });
  txt(fmt(totalGes),         cx2[3] + 4, y + 6, cw2[3] - 8, { bold: true, size: 12, color: GRUEN, align: 'right' });
  y += 24;

  doc.save().rect(LEFT, y - 21 * 2 - 24 - 19 - 22, W, 21 * 2 + 24 + 19 + 22).lineWidth(1).strokeColor(BLAU).stroke().restore();
  y += 10;
});

txt('* Preis pro Stueck inkl. 10 % Aufschlag', LEFT, y, W, { italic: true, size: 7.5, color: '#888888' });
y += 16;

// ============================
// 3. ZUSATZPOSTEN: BAUTENSCHUTZMATTE
// ============================
txt('3.  Zusaetzliche Materialien', LEFT, y, W, { color: BLAU, bold: true, size: 12 });
y += 14;

rect(LEFT, y, W, 22, BLAU);
txt('Bautenschutzmatte', LEFT + 10, y + 6, W - 20, { color: WHITE, bold: true, size: 10 });
y += 22;

rect(LEFT, y, W, 19, HELLBLAU);
['Material', 'Menge', 'Preis (Spanne)', 'Betrag'].forEach((h, i) => {
  txt(h, cx2[i] + 5, y + 5, cw2[i] - 10, { color: BLAU, bold: true, size: 8, align: i === 0 ? 'left' : 'right' });
});
y += 19;

rect(LEFT, y, W, 24, WHITE);
txt('Bautenschutzmatte (Untergrund)', cx2[0] + 5, y + 7, cw2[0] - 10, { size: 9 });
txt('ca. 18-20 m', cx2[1] + 5, y + 7, cw2[1] - 10, { size: 9, align: 'right' });
txt('55 - 70 EUR', cx2[2] + 5, y + 7, cw2[2] - 10, { size: 9, align: 'right' });
txt('55,00 - 70,00 EUR', cx2[3] + 5, y + 7, cw2[3] - 10, { size: 9, align: 'right' });
doc.save().moveTo(LEFT, y + 24).lineTo(LEFT + W, y + 24).lineWidth(0.4).strokeColor('#cccccc').stroke().restore();
y += 24;

doc.save().rect(LEFT, y - 24 - 19 - 22, W, 24 + 19 + 22).lineWidth(1).strokeColor(BLAU).stroke().restore();
y += 14;

// ============================
// 4. GESAMTRANGLISTE
// ============================
doc.save().moveTo(LEFT, y).lineTo(LEFT + W, y).lineWidth(1.5).strokeColor(BLAU).stroke().restore();
y += 10;
txt('4.  Gesamtuebersicht & Rangliste  (180 Stueck + 10 % Preisaufschlag)', LEFT, y, W, { color: BLAU, bold: true, size: 12 });
y += 14;

const sortiert = [...produkte].sort((a, b) => a.preis - b.preis);
const rangFarben = [HELLGRUEN, GRAU, '#fde8e8'];
const rangLabel = ['1.  Guenstigster', '2.  Mittelfeld', '3.  Teuerster'];

const scw = [120, 150, 80, 105, 60];
const scx = [LEFT, LEFT + 120, LEFT + 270, LEFT + 350, LEFT + 455];

rect(LEFT, y, W, 22, BLAU);
['Rang', 'Produkt', 'Preis/Stk.', 'Gesamt 180 Stk.', 'Aufschl.'].forEach((h, i) => {
  txt(h, scx[i] + 5, y + 7, scw[i] - 10, { color: WHITE, bold: true, size: 8.5, align: i < 2 ? 'left' : 'right' });
});
y += 22;

sortiert.forEach((p, i) => {
  const totalGes = STUECK * p.preis * (1 + AUFSCHLAG);
  const g = totalGes.toFixed(2).replace('.', ',') + ' EUR';
  const prS = p.preis.toFixed(2).replace('.', ',') + ' EUR';
  rect(LEFT, y, W, 26, rangFarben[i]);
  txt(rangLabel[i],    scx[0] + 5, y + 8, scw[0] - 10, { bold: true, size: 8.5 });
  txt(p.name,          scx[1] + 5, y + 8, scw[1] - 10, { bold: true, size: 9 });
  txt(prS,             scx[2] + 5, y + 8, scw[2] - 10, { size: 9, align: 'right' });
  txt(g,               scx[3] + 5, y + 7, scw[3] - 10, { bold: true, size: 10, color: GRUEN, align: 'right' });
  txt('+ 10 %',        scx[4] + 5, y + 8, scw[4] - 10, { bold: true, size: 8.5, color: ORANGE_BG.startsWith('#fff') ? '#b36b00' : ROT, align: 'right' });
  doc.save().moveTo(LEFT, y + 26).lineTo(LEFT + W, y + 26).lineWidth(0.4).strokeColor('#cccccc').stroke().restore();
  y += 26;
});
doc.save().rect(LEFT, y - 26 * 3 - 22, W, 26 * 3 + 22).lineWidth(1).strokeColor(BLAU).stroke().restore();

// Bautenschutzmatte in Zusammenfassung
y += 8;
rect(LEFT, y, W, 24, '#fff8e1');
doc.save().moveTo(LEFT, y).lineTo(LEFT + W, y).lineWidth(0.8).strokeColor('#f0a500').stroke().restore();
txt('+ Bautenschutzmatte (ca. 18-20 m)', scx[0] + 5, y + 7, 320, { size: 9, bold: true });
txt('55,00 - 70,00 EUR', scx[3] + 5, y + 7, scw[3] - 10, { size: 9, bold: true, color: '#b36b00', align: 'right' });
doc.save().rect(LEFT, y, W, 24).lineWidth(1).strokeColor('#f0a500').stroke().restore();
y += 24;

// Fusszeile
y += 12;
doc.save().moveTo(LEFT, y).lineTo(LEFT + W, y).lineWidth(0.5).strokeColor('#aaaaaa').stroke().restore();
y += 5;
txt(
  'Hinweis: Alle Produkte gleichen Gefaelle bis 5 % automatisch aus. ' +
  'Preisstand: 10.06.2025. Alle Angaben ohne Gewaehr.',
  LEFT, y, W, { italic: true, size: 7.5, color: '#666666' }
);

doc.end();
console.log('PDF erfolgreich erstellt: Stelzlager_Vergleich.pdf');
